### PR TITLE
Make Reactions extensible

### PR DESCRIPTION
**Pull request by bleistivt**
See https://github.com/hgtonight/Application-Yaga/issues/75 for more information.

### CODE REVIEW

File: controllers/class.reactcontroller.php [hunk 1]
```diff
@@ -49,25 +49,23 @@ public function Index($Type, $ID, $ActionID) {
...
-        throw new Gdn_UserException(T('Yaga.Action.InvalidTargetType'));
-        break;
-    }
+    $Item = null;
+    $AnchorID = '#' . ucfirst($Type) . '_';
+    $ItemOwnerID = 0;
 
-    $Item = $Model->GetID($ID);
+    if(in_array($Type, array('discussion', 'comment', 'activity'))) {
+      $Item = GetRecord($Type, $ID);
```
Comment (hgtonight) [inline controllers/class.reactcontroller.php]: This fails for type `activity`. Looks like this is a 2.2+ feature.

File: controllers/class.reactcontroller.php [hunk 2]
```diff
@@ -49,25 +49,23 @@ public function Index($Type, $ID, $ActionID) {
...
+    else {
+      $this->EventArguments = array(
+        'TargetType' => $Type,
+        'TargetID' => $ID,
+        'Item' => &$Item,
+        'AnchorID' => &$AnchorID,
+        'ItemOwnerID' => &$ItemOwnerID
+      );
+      $this->FireEvent('UnknownType');
+    }
```
Comment (hgtonight) [inline controllers/class.reactcontroller.php]: I would prefer to call the even `CustomType` so the hook is `reactController_customType_handler`.

More importantly there should be a `TypeFound` event argument that plugins should set to true. Then check it right after the event and throw an exception if the type wasn't found.

**Comment by bleistivt** [comment]
I'm not really happy with the `Method` EventArgument anymore.

I'd rather split the ActedModel_Content_Handler into 4 events:

* ActedContent
* TakenContent
* ActionContent
* BestContent

what do you think?

**Comment by hgtonight** [comment]
Splitting the content event out probably makes more sense, but why do you want to switch it?

**Comment by bleistivt** [comment]
Do you mean the event fired from the Get() method?
It could also be called `GetContent` of course.

**Comment by hgtonight** [comment]
I am just wondering **what** made you change your mind?

I am partial to events that make good hook names, e.g. `actedModel_getCustomContent_handler` vs. `actedModel_getContent_handler`.

**Comment by bleistivt** [comment]
I have been using this for a while to integrate yaga reactions in a private plugin:

```php
    public function actedModel_content_handler($sender, &$args) {
        if (!in_array($args['Method'], ['get', 'gettaken']) return;

        $result = Gdn::sql()
            ->select('i.ImageID, i.InsertUserID, i.DateInserted')
            ->from('Image i')
            ->orderBy('i.DateInserted', 'desc')
            ->join('Reaction r', 'i.ImageID = r.ParentID')
            ->where($args['Method'] == 'get' ? 'i.InsertUserID' : 'r.InsertUserID', $args['UserID'])
            ->where('r.ActionID', $args['ActionID'])
            ->where('r.ParentType', 'image')
            ->orderBy('r.DateInserted', 'desc')
            ->get()
            ->result(DATASET_TYPE_ARRAY);

        if ($result) {
            for ($i = 0; $i < count($result); $i++) {
                // Images don't require category based permission checks
                $result[$i]['CategoryID'] = 1;
            }
            $args['CustomSections']['Image'] = $result;
        }
    }
```

This only works nice because it doesn't implement `best` and `action` yet (I simply have to switch out the InsertUserID to switch between `taken` and `gettaken`).
These would need entirely different queries and the function would turn out to be very long.

It would essentially be a:

    switch (args['Method']) {
         ...

which is usually a sign that a function should be split into multiple smaller ones.

So you would call them like this?

* GetCustomContent
* TakenCustomContent
* ActionCustomContent
* BestCustomContent

**Comment by hgtonight** [comment]
@bleistivt Very insightful.

I am partial to `GetCustom`, `GetCustomTaken`, `GetCustomAction`, `GetCustomRecord`, etc.

**Comment by hgtonight** [comment]
Thanks for all your help!

**Pull request merged**